### PR TITLE
Fixed #24128 -- Made admindocs' TemplateDetailView account for template loaders.

### PR DESCRIPTION
**Pull request by krisavi**
Modification for TemplateDetailView in admindocs to go through the directories from template loaders.

https://code.djangoproject.com/ticket/24128

Might need some improvements.

### CODE REVIEW

**Comment by timgraham** [comment]
Some tests are needed.

**Comment by krisavi** [comment]
I am thinking how that test should be written, Probably can add to one of the views some template tag and then do a test towards TemplateDetailView to check if folder from app_dir was checked?

To be honest, I have not written tests (which I know is wrong way to do some things). Will try to figure it out.

**Comment by krisavi** [comment]
I noticed a "bug" I made, in case of cached loaders I am not going through them properly. I made a fix for it but I don't consider it elegant, it works, but probably needs some refining.

At the same time I noticed the tests are only using cached loaders, that means there aren't tests for cases when the cached template loader isn't used.

The test written, went through cached loaders quite fine and passed. On my local development environment where the issue came up I don't have cached template loaders enabled and it works, tried also enabling caching and no problems.

As I can also see... I can't use git correctly

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by krisavi** [comment]
I was fixing the issue with tests failed and noticed it is possible to make workaround with multiple template engines as well, the bug #24125. It was actually even implemented in my initial code to go through all the engines.

Not sure if I should commit it into this pull request or not. The test written works with both solutions but will not pass AdminDocViewWithMultipleEngines test without the fix for #24125. Not sure why "test_template_detail" passes (was in the initial code already), it uses the same function and fails at the same location.

File: django/contrib/admindocs/views.py [hunk 1]
```diff
@@ -353,8 +355,14 @@ def get_context_data(self, **kwargs):
             # Non-trivial TEMPLATES settings aren't supported (#24125).
             pass
         else:
-            # This doesn't account for template loaders (#24128).
-            for index, directory in enumerate(default_engine.dirs):
+            directories = set(default_engine.dirs)
+            for loader in default_engine.template_loaders:
+                if isinstance(loader, Loader):
+                    directories.update(loader.get_dirs())
+                elif isinstance(loader, Cached_Loader):
```
Comment (timgraham) [inline django/contrib/admindocs/views.py]: Perhaps the `Cached_Loader` should have a `get_dirs()` method? Special casing it here doesn't look like a good separation of concerns (admindocs knowing about internals of the template system).

Comment (krisavi) [inline django/contrib/admindocs/views.py]: I agree, when I was making that change I was thinking about it, but have to look into the Cached_Loader, maybe easiest would be to add the get_dirs(), the test should still work and then there is no need to check which object type it is.

I can look into it tomorrow to see if I am able to write the change.

**Comment by krisavi** [comment]
Since there are multiple loaders in cached loaders and they may have different ways to get dirs then I saw this as being the best solution for that problem. It will iterate over them and ask to get the get_dir functions of the specific loader.

File: django/template/loaders/cached.py [hunk 2]
```diff
@@ -17,6 +17,13 @@
 
 class Loader(BaseLoader):
 
+    def get_dirs(self):
+        dirs = set()
+        for loader in self.loaders:
+            if loader is not None:
+                dirs.update(loader.get_dirs())
+        return dirs
+
```
Comment (pope1ni) [inline django/template/loaders/cached.py]: Reduce to 20% of the lines and avoid multiple updates to the set:

```python
return {loader.get_dirs() for loader in self.loaders if loader is not None}
```

Comment (krisavi) [inline django/template/loaders/cached.py]: I thought I will eventually get to that kind of answer on the PR. Was thinking about some similar solution, but wasn't able to write it, then I pushed something that I knew was working.

**Comment by timgraham** [comment]
Are you aware of the test failures? You can squash commits when updating.

**Comment by pope1ni** [comment]
@timgraham, @krisavi 

This is my fault - i mistook `set.update()` for `set.add()` and gave a duff suggestion. Either revert the last commit added or change the set comprehension to `set(itertools.chain.from_iterable(...))`

I just had another look at this and the changes make use of sets. Is ordering important here? If so, should an ordered set be used in both areas that have been changed?

**Comment by krisavi** [comment]
@pope1ni 

The ordering is not important, but if I am right then the use of set not having duplicates is useful here. 

It will check all the dirs the template could be reported and will then report if it is found in any of them. Duplicates would mean duplicate checks as well, which are not needed. Ordering is not relevant.

**Comment by timgraham** [comment]
I think ordering is important -- we want to show the same ordering that the template loader uses. Removing duplicates isn't important.

**Comment by timgraham** [comment]
Kris, do you want to try to complete this?

**Comment by krisavi** [comment]
I will look into it again. I think there has been some changes and might needs some changes.

**Comment by timgraham** [comment]
Closing due to inactivity.

File: tests/admin_docs/test_views.py [hunk 3]
```diff
@@ -156,6 +160,12 @@ def test_templatetag_index(self):
         response = self.client.get(reverse('django-admindocs-tags'))
         self.assertContains(response, '<title>Template tags</title>', html=True)
 
+    def test_template_detail_loader(self):
+        # Overridden because non-trivial TEMPLATES settings aren't supported
+        # but the page shouldn't crash (#24125).
+        response = self.client.get(reverse('django-admindocs-templates', args=['view_for_loader_test.html']))
+        self.assertContains(response, '<h1>Template: "view_for_loader_test.html"</h1>', html=True)
+
```
Comment (felixxm) [inline tests/admin_docs/test_views.py:168]: There is no need to override this test.

**Pull request closed without merging**